README.md
kntnr
=====
[![Travis-CI Build Status](https://travis-ci.org/yutannihilation/kntnr.svg?branch=master)](https://travis-ci.org/yutannihilation/kntnr)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/yutannihilation/kntnr?branch=master&svg=true)](https://ci.appveyor.com/project/yutannihilation/kntnr)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/kntnr)](https://cran.r-project.org/package=kntnr)

Retrieve data from [kintone](https://www.kintone.com/) via its API. kintone is an enterprise application platform.

## Installation

Stable CRAN version

```r
install.packages("kntnr")
```

Dev version

```r
devtools::install_github("yutannihilation/kntnr")
```

## Usage

### Authentication

To use kintone API, kntnr needs the following environmental variables.

* `KNTN_URL`: kintone URL
* `KNTN_AUTH_TYPE`: authentication type (password or token)
* `KNTN_AUTH`: API token or BASE64 encoded login name and password

`kntn_set_auth()` sets these environmental variables interactively.

```r
library(kntnr)

# password-based authentication
kntn_set_auth()

# token-based authentication
kntn_set_auth(auth_type = "token")
```

To set another information, run `kntn_set_auth()` with `overwrite = TRUE`.

```r
kntn_set_auth(overwrite = TRUE)
```

Or, unset the environmental variables by `kntn_unset_auth()`.

```r
kntn_unset_auth()
```

#### For non-interactive use

You can define the environmental variables in `.Renviron`, which is loaded at the start of an R session. For more information, see `?Startup`.

```r
KNTN_URL = https://example.kintone.com/
KNTN_AUTH_TYPE = token
KNTN_AUTH = 1234567890
```

### Record API

`kntn_record()` gets a single record from the specified kintone application. `kntn_records()` retrieves multiple records at once. If the number of records is more than records_per_request (the default is 100), `kntn_records()` automatically splits the request into smaller subrequests.

```r
app <- 10

# get a single record
d <- kntn_record(app, id = 1)

# get records up to 1000 (default)
d <- kntn_records(app)

# get records up to 5000 records at the latency of 500 records/request.
d <- kntn_records(app, max_records = 5000, records_per_request = 500L)
```

Some types like SUBTABLE are converted as nested data.frame.
You can unnest them by using `kntn_unnest()`.

```r
kntn_unnest(d)
```

### File API

`kntn_file()` gets a file from kintone API and try to parse it with `httr::content()`. If you want to parse it by yourself, specify `as = "raw"` or `as = "text"`. 

```r
d <- kntn_record(app, id = 1)

f <- kntn_file(fileKey = x$Attachment[[1]]$fileKey[1])
```


NEWS.md
# kntnr (development version)

# kntnr 0.4.4

* Fix errors on the CRAN check.

# kntnr 0.4.3

* Fix compatibility problem with dplyr v1.0.0.

# kntnr 0.4.2

* Fix compatibility problem with tidyr v1.0.0.

# kntnr 0.4.1

## Breaking changes

* `kntn_file()` loses `app` argument, as app ID is not needed for File API.

## Minor changes

* Improve performance.
* Fix compatibility problems with dplyr >0.7.

# kntnr 0.4.0

* First release.


tests/testthat/test-calc_ranges.R
context("calc_ranges")

test_that("calc ranges without a fraction", {
  expect_equal(
    calc_ranges(
      first_offset = 0L,
      max_records = 1000L,
      records_per_request = 100L
    ),
    tibble::tibble(
      offsets = c(  0L, 100L, 200L, 300L, 400L, 500L, 600L, 700L, 800L, 900L),
      limits  = c(100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L)
    )
  )
})

test_that("calc ranges with one length short", {
  expect_equal(
    calc_ranges(
      first_offset = 0L,
      max_records = 999L,
      records_per_request = 100L
    ),
    tibble::tibble(
      offsets = c(  0L, 100L, 200L, 300L, 400L, 500L, 600L, 700L, 800L, 900L),
      limits  = c(100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L,  99L)
    )
  )
})

test_that("calc ranges with one length over", {
  expect_equal(
    calc_ranges(
      first_offset = 0L,
      max_records = 1001L,
      records_per_request = 100L
    ),
    tibble::tibble(
      offsets = c(  0L, 100L, 200L, 300L, 400L, 500L, 600L, 700L, 800L, 900L, 1000L),
      limits  = c(100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L,    1L)
    )
  )
})

test_that("calc ranges without a fraction and with offset", {
  expect_equal(
    calc_ranges(
      first_offset = 1L,
      max_records = 1000L,
      records_per_request = 100L
    ),
    tibble::tibble(
      offsets = c(  1L, 101L, 201L, 301L, 401L, 501L, 601L, 701L, 801L, 901L),
      limits  = c(100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L, 100L)
    )
  )
})

test_that("calc ranges without a fraction and with offset", {
  expect_equal(
    calc_ranges(
      first_offset =1001L,
      max_records = 1000L,
      records_per_request = 100L
    ),
    tibble::tibble(
      offsets = c(1001L, 1101L, 1201L, 1301L, 1401L, 1501L, 1601L, 1701L, 1801L, 1901L),
      limits  = c( 100L,  100L,  100L,  100L,  100L,  100L,  100L,  100L,  100L,  100L)
    )
  )
})


tests/testthat/test-parse-field.R
context("parse")

test_parse_field <- function(txt, expect, skip_on_windows = FALSE) {
  if (skip_on_windows) {
    skip_on_os("windows")
  }

  j <- jsonlite::fromJSON(txt, simplifyVector = FALSE)
  test_that(sprintf("parsing %s works", j$type), {
    x <- kntn_parse_records(list(list(test = j)))
    if(is.function(expect)){
      expect_true(expect(x$test))
    } else {
      expect_equivalent(x$test, expect)
    }
  })
}

test_parse_field(
  '{
    "type": "RECORD_NUMBER",
    "value": "7"
  }',
  "7"
)

test_parse_field(
  '{
    "type": "__ID__",
    "value": "3"
  }',
  3L
)

test_parse_field(
  '{
    "type": "__REVISION__",
  "value": "9"
  }',
  9L
)

test_parse_field(
  '{
    "type": "CREATOR",
    "value": {
        "code": "john-d",
        "name": "John Doe"
    }
  }',
  "john-d"
)

test_parse_field(
  '{
  "type": "CREATOR",
  "value": null
  }',
  NA_character_
)

test_parse_field(
  '{
    "type": "CREATED_TIME",
    "value": "2015-01-22T15:07:00Z"
  }',
  as.POSIXct("2015-01-22 15:07:00", tz = "UTC"),
  skip_on_windows = TRUE
)

test_parse_field(
  '{
  "type": "CREATED_TIME",
  "value": null
  }',
  function(x) is.na(x) && lubridate::is.POSIXct(x)
)

test_parse_field(
  '{
        "type": "MODIFIER",
        "value": {
            "code": "john-d",
            "name": "John Doe"
        }
  }',
  "john-d"
)

test_parse_field(
  '{
    "type": "UPDATED_TIME",
    "value": "2015-01-22T15:07:00Z"
  }',
  as.POSIXct("2015-01-22 15:07:00", tz = "UTC"),
  skip_on_windows = TRUE
)

test_parse_field(
  '{
    "type": "SINGLE_LINE_TEXT",
    "value": "Hello kintone"
  }',
  "Hello kintone"
)

test_parse_field(
  '{
    "type": "NUMBER",
    "value": "777"
  }',
  777
)

test_parse_field(
  '{
    "type": "CALC",
    "value": "777888999"
  }',
  "777888999"
)

test_parse_field(
  '{
    "type": "MULTI_LINE_TEXT",
    "value": "Hello\\nkintone"
  }',
  "Hello\nkintone"
)

test_parse_field(
  '{
    "type": "RICH_TEXT",
    "value": "<a href=\\"https://www.kintone.com\\">kintone</a>"
  }',
  '<a href="https://www.kintone.com">kintone</a>'
)

test_parse_field(
  '{
    "type": "CHECK_BOX",
    "value": [
    "Choice 1",
    "Choice 2"
    ]
  }',
  list(c("Choice 1", "Choice 2"))
)

test_parse_field(
  '{
  "type": "CHECK_BOX",
  "value": [
  ]
  }',
  list(character(0))
)

test_parse_field(
  '{
    "type": "RADIO_BUTTON",
    "value": "Choice 3"
  }',
  "Choice 3"
)

test_parse_field(
  '{
    "type": "DROP_DOWN",
    "value": "Choice 1"
  }',
  "Choice 1"
)

test_parse_field(
  '{
    "type": "MULTI_SELECT",
    "value": [
    "Choice 1",
    "Choice 2"
    ]
  }',
  list(c("Choice 1", "Choice 2"))
)

test_parse_field(
  '{
    "type": "FILE",
    "value": [
        {
        "contentType": "text/plain",
        "fileKey":"20150417022053715283FF97DC413CBC4B7A41C",
        "name": "kintoneUpdates.txt",
        "size": "25302"
        },
        {
        "contentType": "text/plain",
        "fileKey": "20150417022159ECFC1223C4B34C939E1B9BD25",
        "name": "APIUpdatestxt",
        "size": "20311"
        }
    ]
  }',
  function(x) {
    is.list(x) && dplyr::is.tbl(x[[1]]) &&
      identical(colnames(x[[1]]), c("contentType", "fileKey", "name", "size"))
  }
)

test_parse_field(
  '{
  "type": "FILE",
  "value": []
  }',
  list(tibble::tibble())
)


test_parse_field(
  '{
    "type": "LINK",
    "value": "https://www.kintone.com/"
  }',
  "https://www.kintone.com/"
)

test_parse_field(
  '{
    "type": "DATE",
    "value": "2015-04-15"
  }',
  as.Date("2015-04-15")
)

test_parse_field(
  '{
  "type": "DATE",
  "value": null
  }',
  function(x) is.na(x) && lubridate::is.Date(x)
)

test_parse_field(
  '{
    "type": "TIME",
    "value": "09:00"
  }',
  "09:00"
)

test_parse_field(
  '{
    "type": "DATETIME",
    "value": "2015-03-17T10:20:00Z"
  }',
  as.POSIXct("2015-03-17 10:20:00", tz = "UTC"),
  skip_on_windows = TRUE
)


test_parse_field(
  '{
    "type": "USER_SELECT",
    "value": [
        {
        "code": "john-d",
        "name": "John Doe"
        },
        {
        "code": "jane-r",
        "name": "Jane Roe"
        }
    ]
  }',
  list(c("john-d", "jane-r"))
)

test_parse_field(
  '{
    "type": "USER_SELECT",
    "value": []
  }',
  list(character(0))
)

test_parse_field(
  '{
    "type": "ORGANIZATION_SELECT",
    "value": [
        {
            "code": "human_resources",
            "name": "Human Resources"
        },
        {
            "code": "sales",
            "name": "Sales"
        }
    ]
  }',
  list(c("human_resources", "sales"))
)

test_parse_field(
  '{
    "type": "GROUP_SELECT",
    "value": [
        {
            "code": "project_manager",
            "name": "Project Manager"
        },
        {
            "code": "team_leader",
            "name": "Team Leader"
        }
    ]
  }',
  list(c("project_manager", "team_leader"))
)

test_parse_field(
  '{
    "type": "CATEGORY",
    "value": [
    "category1",
    "category2"
    ]
  }',
  list(c("category1", "category2"))
)

test_parse_field(
  '{
    "type": "STATUS",
    "value": "Not started"
  }',
  "Not started"
)

test_parse_field(
  '{
    "type": "STATUS_ASSIGNEE",
    "value": [
        {
        "code": "john-d",
        "name": "John Doe"
        }
    ]
  }',
  list(c("john-d"))
)

test_parse_field(
  '{
    "type": "SINGLE_LINE_TEXT",
    "value": "Code003"
  }',
  "Code003"
)

test_parse_field(
  '{
    "type": "SUBTABLE",
    "value": [
        {
            "id": "48277",
            "value": {
                "textfield_0": {
                    "type": "SINGLE_LINE_TEXT",
                    "value": "Hello kintone 1"
                },
                "numberfield_0": {
                    "type": "NUMBER",
                    "value": "5"
                },
                "checkboxfield_0": {
                    "type": "CHECK_BOX",
                    "value": [
                    "Choice1"
                    ]
                }
            }
        },
        {
            "id": "48278",
            "value": {
                "textfield_0": {
                    "type": "SINGLE_LINE_TEXT",
                    "value": "Hello kintone 2"
                },
                "numberfield_0": {
                    "type": "NUMBER",
                    "value": "7"
                },
                "checkboxfield_0": {
                    "type": "CHECK_BOX",
                    "value": [
                    "Choice2"
                    ]
                }
            }
        }
    ]
  }',
  function(x) {
    is.list(x) && dplyr::is.tbl(x[[1]]) &&
      identical(colnames(x[[1]]), c("textfield_0", "numberfield_0", "checkboxfield_0"))
  }
)

test_parse_field(
  '{
    "type": "SUBTABLE",
    "value": []
  }',
  list(tibble::tibble())
)


test_that('kntn_parse_subtable_one() works',
          {
            d <- kntn_parse_subtable_one(
              jsonlite::fromJSON('[
  {
    "id": "48277",
    "value": {
      "textfield_0": {
        "type": "SINGLE_LINE_TEXT",
        "value": "Hello kintone 1"
      },
      "numberfield_0": {
        "type": "NUMBER",
        "value": "5"
      },
      "checkboxfield_0": {
      "type": "CHECK_BOX",
        "value": [
          "Choice1"
        ]
      }
    }
  },
  {
    "id": "48278",
    "value": {
      "textfield_0": {
        "type": "SINGLE_LINE_TEXT",
        "value": "Hello kintone 2"
      },
      "numberfield_0": {
        "type": "NUMBER",
        "value": "7"
      },
      "checkboxfield_0": {
        "type": "CHECK_BOX",
        "value": [
          "Choice2"
        ]
      }
    }
  }
]', simplifyVector = FALSE))
            expect_true(dplyr::is.tbl(d))})


tests/testthat/test-parse-record.R
context("kntn_parse and kntn_unnest")

library(dplyr, warn.conflicts = FALSE)

test_that("parsing a single record works", {
  record_file <- system.file("extdata/record.json", package = "kntnr")
  record_obj  <- jsonlite::fromJSON(record_file, simplifyVector = FALSE)$record %>%
    kntn_parse_record

  expect_identical(dim(record_obj), c(1L, 21L))
})

test_that("parsing multiple records works", {
  records_file <- system.file("extdata/records.json", package = "kntnr")
  records_obj  <- jsonlite::fromJSON(records_file, simplifyVector = FALSE)$records %>%
    kntn_parse_records

  expect_identical(dim(records_obj), c(3L, 6L))
})

test_that("unnesting records works", {
  records_file <- system.file("extdata/records.json", package = "kntnr")
  records_obj  <- jsonlite::fromJSON(records_file, simplifyVector = FALSE)$records %>%
    kntn_parse_records %>%
    kntn_unnest

  expect_identical(dim(records_obj), c(6L, 8L))
  expect_equivalent(table(records_obj$record_id),
                    as.table(purrr::set_names(c(3,1,2), 1:3)))

  na_count <- dplyr::group_by(records_obj, record_id) %>%
    dplyr::summarize(na_count_checkbox = sum(is.na(checkbox)),
                     na_count_subfield1 = sum(is.na(subfield1)))
  expect_identical(na_count$na_count_checkbox, c(3L, 0L, 0L))
  expect_identical(na_count$na_count_subfield1, c(0L, 0L, 2L))

})


test_that("unnesting terriblly-nested records works", {
  records_file <- system.file("extdata/records_nested.json", package = "kntnr")
  records_obj  <- jsonlite::fromJSON(records_file, simplifyVector = FALSE)$records %>%
    kntn_parse_records %>%
    kntn_unnest

  expect_identical(dim(records_obj), c(5L, 8L))
  expect_equivalent(table(records_obj$record_id),
                    as.table(purrr::set_names(c(3,2), 1:2)))

  na_count <- dplyr::group_by(records_obj, record_id) %>%
    dplyr::summarize(na_count_fileKey = sum(is.na(fileKey)),
                     na_count_subfield1 = sum(is.na(subfield1)))
  expect_identical(na_count$na_count_fileKey, c(0L, 2L))
  expect_identical(na_count$na_count_subfield1, c(0L, 1L))

})


R/util.R
# util functions

guess_url <- function(url_str) {
  if(!purrr::is_scalar_character(url_str)) {
    stop("url_str must be character and the length is 1")
  }

  if(!stringr::str_detect(url_str, "^http")){
    url_str <- paste0("https://", url_str)
  }

  url <- httr::parse_url(url_str)

  url$scheme <- "https"
  url$path <- NULL

  httr::build_url(url)
}


calc_ranges <- function(first_offset = 0L,
                        max_records = 1000L,
                        records_per_request = 100L) {
  ranges <- list()

  ranges$offsets <- seq(from = first_offset,
                        to = first_offset + max_records,
                        by = records_per_request)

  ranges$limits <- rep(records_per_request, length(ranges$offsets) - 1)

  # treat a fraction
  fraction <- max_records %% records_per_request
  if(fraction == 0) {
    # remove the range of zero-length
    length(ranges$offsets) <- length(ranges$offsets) - 1
  } else {
    ranges$limits <- append(ranges$limits, fraction)
  }

  tibble::as_tibble(ranges)
}


R/kntn_auth.R
#' Authorization Header for 'kintone'
#'
#' \code{kntn_get_authorization_header()} constructs a header for authorization using \link[httr]{add_headers}.
#' \code{kntn_set_auth()} interactivery asks for the type of authentication and the credential and store
#' them as environmental variables \code{KNTN_URL}, \code{KNTN_AUTH_TYPE} and \code{KNTN_AUTH}.
#' If you want to avoid interaction, please manually set these.
#' \code{kntn_unset_auth()} unsets these environmental variables.
#'
#' @name kntn_auth
#'
#' @param auth_type
#'   Type of Authentication. \code{password} uses username and password.
#'   \code{token} uses per-app token.
#' @param overwrite
#'   If \code{TRUE}, overwrite the existing environmental variables.
#'
#' @seealso \url{https://developer.kintone.io/hc/en-us/articles/212495188/#userauthentication}
#'
#' @examples
#' \dontrun{
#' # set KNTN_URL, KNTN_AUTH and KNTN_AUTH_TYPE interactively.
#' # By default, auth_type is password.
#' kntn_set_auth()
#'
#' Sys.getenv(c("KNTN_URL", "KNTN_AUTH", "KNTN_AUTH_TYPE"))
#'
#' # This will return `X-Cybozu-Authorization` header.
#' kntn_get_authorization_header()
#'
#' # Clear environmental variables before trying to use another set of authorization info.
#' kntn_unset_auth()
#' kntn_set_auth(auth_type = "token")
#'
#' # This will return `X-Cybozu-API-Token` header.
#' kntn_get_authorization_header()
#'
#' # To avoid interaction, set these environmental variables manually.
#' Sys.setenv("KNTN_URL" = "https://example.cybozu.com/")
#' Sys.setenv("KNTN_AUTH" = "abcdefg")
#' Sys.setenv("KNTN_AUTH_TYPE" = "token")
#' }
#' @export
kntn_get_authorization_header <- function() {
  if(!kntn_check_envvars_defined()) {
    stop("Please set necessary environmental variables by .Renviron or kntn_set_auth()")
  }

  auth <- Sys.getenv("KNTN_AUTH")
  auth_type <- Sys.getenv("KNTN_AUTH_TYPE")

  switch(auth_type,
         "password" = return(httr::add_headers(`X-Cybozu-Authorization` = auth)),
         "token"    = return(httr::add_headers(`X-Cybozu-API-Token` = auth)),
         stop(sprintf("Unknown KNTN_AUTH_TYPE: %s", auth_type)))
}

#' @rdname kntn_auth
#' @export
kntn_set_auth <- function(auth_type = c("password", "token"), overwrite = FALSE) {
  if (!interactive()) {
    stop("If you want to run kntnr in noninteractive environment,\n",
         "please manually set these environmental variables:\n",
         "  - KNTN_URL\n",
         "  - KNTN_AUTH\n",
         "  - KNTN_AUTH_TYPE\n")
  }

  if (kntn_check_envvars_defined(assume_all = FALSE) && !overwrite) {
    warning("At least one of KNTN_URL, KNTN_AUTH and KNTN_AUTH_TYPE is defined.\n",
            "To overrite these, please rerun with 'overwrite = TRUE'\n")
    return()
  }

  auth_type <- match.arg(auth_type)

  url <- ask_for_nonsecret("kintone URL (e.g. https://example.cybozu.com/): ")

  if(identical(auth_type, "password")) {
    username <- ask_for_secret("username for kintone: ")
    password <- ask_for_secret("password for kintone: ")
    auth <- base64enc::base64encode(charToRaw(paste(username, password, sep = ":")))
  } else {
    auth <- ask_for_secret("API token for kintone: ")
  }

  Sys.setenv("KNTN_URL" = url)
  Sys.setenv("KNTN_AUTH" = auth)
  Sys.setenv("KNTN_AUTH_TYPE" = auth_type)

  message("Environmental variables KNTN_URL, KNTN_AUTH and KNTN_AUTH_TYPE are set.")
  message("To unset these, please run kntn_unset_auth().")
}

kntn_check_envvars_defined <- function(assume_all = TRUE) {
  envvars_defined <- c("KNTN_URL", "KNTN_AUTH", "KNTN_AUTH_TYPE") %in% names(Sys.getenv())
  if(assume_all) {
    all(envvars_defined)
  } else {
    any(envvars_defined)
  }
}


#' @rdname kntn_auth
#' @export
kntn_unset_auth <- function() {
  Sys.unsetenv("KNTN_URL")
  Sys.unsetenv("KNTN_AUTH")
  Sys.unsetenv("KNTN_AUTH_TYPE")
}

ask_for_nonsecret <- function(prompt) {
  cat(prompt)
  readLines(n = 1L)
}

ask_for_secret <- function(prompt) {
  if(rstudioapi::isAvailable()) {
    rstudioapi::askForPassword(prompt)
  } else {
    # Fallback to ask_for_nonsecret
    ask_for_nonsecret(prompt)
  }
}


R/kntn_file.R
#' 'kintone' File API
#'
#' Get a file from kintone API and parse it with \link[httr]{content}. If you want to parse it by yourself,
#' specify \code{as = "raw"} or \code{as = "text"}.
#'
#' @name kntn_file
#'
#' @seealso \url{https://developer.kintone.io/hc/en-us/articles/212494468/}
#'
#' @param fileKey File key.
#' @param verbose If \code{TRUE}, output logs verbosely.
#' @param as \code{as} parameter passed to \link[httr]{content}.
#' @param type \code{type} parameter passed to \link[httr]{content}.
#' @param encoding \code{encoding} parameter passed to \link[httr]{content}.
#'
#' @examples
#' \dontrun{
#' kntn_set_auth()
#' app <- 10
#'
#' # get a single record with a file attachment field
#' d <- kntn_record(app, id = 1)
#'
#' f <- kntn_file(fileKey = x$Attachment[[1]]$fileKey[1])
#' }
#'
#' @export
kntn_file <- function(fileKey, verbose = FALSE, as = NULL, type = NULL, encoding = NULL) {
  query_params <- list(`fileKey`  = fileKey)
  if(!purrr::is_scalar_character(fileKey)) {
    stop("fileKey must be a character scalar.")
  }

  if(!kntn_check_envvars_defined()) {
    stop("Please set necessary environmental variables by .Renviron or kntn_set_auth()")
  }

  url <- guess_url(Sys.getenv("KNTN_URL"))

  res <- kntn_api(
    "GET",
    url = url,
    path = "/k/v1/file.json",
    query = query_params,
    verbose = verbose
  )

  httr::content(res, as = as, type = type, encoding = encoding)
}


R/kntn_unnest.R
#' Unnest 'kintone' Records
#'
#' The data retrieved by \link{kntn_records} may contain nested data.frames.
#' \code{kntn_unnest()} unnests them by using \link[tidyr]{unnest}.
#' Note that this function is very experimental and may not work well for all data.
#'
#' @param records Data.frame retrieved by kntn_records()
#'
#' @export
kntn_unnest <- function(records) {
  # data.frame
  nested_df_cols <- purrr::map_lgl(records, is_nested_df)
  nested_df_colnames <- names(nested_df_cols)[nested_df_cols]

  try_unnest_recursively <- FALSE

  if (length(nested_df_colnames) > 0) {
    # SUBTABLE may contain nested fields
    try_unnest_recursively <- TRUE

    records <- dplyr::mutate_at(records, nested_df_colnames, fill_dummy_df)
  }

  # character
  nested_chr_cols <- purrr::map_lgl(records, is_nested_chr)
  nested_chr_colnames <- names(nested_chr_cols)[nested_chr_cols]

  if (length(nested_chr_colnames) > 0) {
    records <- dplyr::mutate_at(records, nested_chr_colnames, fill_dummy_chr)
  }

  # We have to unnest one by one, otherwise we will see the error:
  # "All nested columns must have the same number of elements."
  unnest_legacy <- get_unnest_legacy()
  for (col in c(nested_df_colnames, nested_chr_colnames)) {
    records <- unnest_legacy(records, !!rlang::sym(col), .drop = FALSE)
  }

  if (try_unnest_recursively) {
    records <- kntn_unnest(records)
  }

  records
}

# deal with the breaking change introduced in tidyr v1.0.0
# c.f. https://tidyr.tidyverse.org/dev/articles/in-packages.html
get_unnest_legacy <- function() {
  if (utils::packageVersion("tidyr") > "0.8.99") {
    # Using tidyr::unnest_legacy directly will cause CRAN check NOTE
    utils::getFromNamespace("unnest_legacy", "tidyr")
  } else {
    tidyr::unnest
  }
}


is_nested_df <- function(x) { is.list(x) && dplyr::is.tbl(x[[1]])}
is_nested_chr <- function(x) { is.list(x) && is.character(x[[1]])}

fill_dummy_chr <- function(x, nm) {
  purrr::map_if(x, ~ length(.) == 0, ~ NA_character_)
}

fill_dummy_df <- function(x, nm) {
  idx_empty <- purrr::map_int(x, nrow) == 0
  idx_nonempty <- !idx_empty

  if(any(idx_nonempty)) {
    dummy_df_example <- x[idx_nonempty][[1]]
    dummy_colnames <- colnames(dummy_df_example)
    dummy <- purrr::map(dummy_df_example,
                        ~ purrr::when(.,
                                    is_nested_chr(.) ~ list(character(0)),
                                    is_nested_df(.) ~ list(tibble::tibble()),
                                    ~ NA))
  } else {
    dummy_colnames <- nm
    dummy <- purrr::rerun(length(dummy_colnames), NA)
  }

  names(dummy) <- dummy_colnames
  dummy_df <- tibble::as_tibble(dummy)

  purrr::map_if(x, idx_empty, ~ dummy_df)
}


R/kntn_record.R
#' 'kintone' Record API
#'
#' \code{kntn_record()} gets a single record from the specified kintone application.
#' \code{kntn_records()} retrieves multiple records at once. If the number of records is more than
#' \code{records_per_request} (the default is 100), \code{kntn_records()} automatically splits the
#' request into smaller subrequests.
#'
#' @name kntn_record
#'
#' @param app App ID.
#' @param id Record ID.
#' @param query Query (e.g. \code{"updated_time > \\"2012-02-03T09:00:00+0900\\" and updated_time < \\"2012-02-03T10:00:00+0900\\""}).
#' @param fields Names of fields.
#' @param verbose If \code{TRUE}, output logs verbosely.
#' @param as Desired type of output: \code{data.frame}, \code{list} or \code{text} (\code{data.frame} means \link[dplyr]{tbl_df}).
#' @param max_records Max number of records to get.
#' @param offset Offset of records.
#' @param records_per_request Number of records per request (max: 100).
#'
#' @details
#' A field will be converted to the correspondent object by the type:
#' \itemize{
#' \item RECORD_NUMBER: \code{character}
#' \item __ID__: \code{integer}
#' \item __REVISION__: \code{integer}
#' \item CREATOR: \code{character} (code)
#' \item CREATED_TIME: \code{POSIXct}
#' \item MODIFIER: \code{character} (code)
#' \item UPDATED_TIME: \code{POSIXct}
#' \item SINGLE_LINE_TEXT: \code{character}
#' \item NUMBER: \code{numeric}
#' \item CALC: \code{character}
#' \item MULTI_LINE_TEXT: \code{character}
#' \item RICH_TEXT: \code{character}
#' \item CHECK_BOX: nested \code{character}
#' \item RADIO_BUTTON: \code{character}
#' \item DROP_DOWN: \code{character}
#' \item MULTI_SELECT: nested \code{character}
#' \item FILE: nested \code{\link[dplyr]{tbl_df}}
#' \item LINK: \code{character}
#' \item DATE: \code{Date}
#' \item TIME: \code{character} (R has no correspondent class for this)
#' \item DATETIME: \code{POSIXct}
#' \item USER_SELECT: nested \code{character} (code)
#' \item ORGANIZATION_SELECT: nested \code{character} (code)
#' \item GROUP_SELECT: nested \code{character} (code)
#' \item CATEGORY: nested \code{character}
#' \item STATUS: \code{character}
#' \item STATUS_ASSIGNEE: \code{character}
#' \item SUBTABLE: nested \code{tbl}
#' }
#'
#' Some types will be converted to nested objects. You can unnest these fields by
#' \link{kntn_unnest}.
#'
#' @examples
#' \dontrun{
#' kntn_set_auth()
#'
#' app <- 10
#'
#' # get a single record
#' d <- kntn_record(app, id = 1)
#'
#' # get records up to 1000 (default)
#' d <- kntn_records(app)
#'
#' # get records up to 5000 records at the latency of 500 records/request.
#' d <- kntn_records(app, max_records = 5000, records_per_request = 500L)
#'
#' # get records as list
#' d <- kntn_records(app, as = "list")
#'
#' # get records matched with the specified query and fields.
#' # See https://developer.kintone.io/hc/en-us/articles/213149287/ for the query syntax
#' d <- kntn_records(app, fields = c("timestamp", "value"),
#'                   query = "updated_time > \"2016-10-03T09:00:00+0900\"")
#'
#' # Some types like SUBTABLE are converted as nested data.frame.
#' # You can unnest them by using kntn_unnest.
#' kntn_unnest(d)
#' }
#'
#' @seealso \url{https://developer.kintone.io/hc/en-us/articles/213149287/}
NULL

kntn_api <- function(verb, url, path, app, query, verbose = FALSE) {
  if(verbose) message("query is:", query)

  res <- httr::VERB(
    verb   = verb,
    url    = url,
    path   = path,
    config = kntn_get_authorization_header(),
    query = query
  )

  kntn_stop_for_status(res)

  res
}

#' @rdname kntn_record
#' @export
kntn_record <- function(app, id, as = c("data.frame", "list", "text"), verbose = FALSE) {
  if(!kntn_check_envvars_defined()) {
    stop("Please set necessary environmental variables by .Renviron or kntn_set_auth()")
  }

  url <- guess_url(Sys.getenv("KNTN_URL"))

  as <- match.arg(as)

  query_params <- list(`app` = app,
                       `id`  = id)

  res <- kntn_api(
    "GET",
    url = url,
    path = "/k/v1/record.json",
    query = query_params,
    verbose = verbose
  )

  record_text <- httr::content(res, as = "text")

  if (as == "text") {
    return(record_text)
  }

  record_json <- jsonlite::fromJSON(record_text, simplifyVector = FALSE)

  if (as == "list") {
    return(record_json)
  }

  kntn_parse_record(record_json$record)
}


kntn_records_once <- function(url, app, fields, query,
                              as = c("data.frame", "list", "text"), verbose = FALSE) {

  as <- match.arg(as)

  query_params <- list(`app` = app,
                       fields = paste0(fields, collapse = ","),
                       query = query) %>%
    purrr::discard(is.null)

  res <- kntn_api(
    "GET",
    url = url,
    path = "/k/v1/records.json",
    query = query_params,
    verbose = verbose
  )

  records_text <- httr::content(res, as = "text")

  if (as == "text") {
    return(records_text)
  }

  records_json <- jsonlite::fromJSON(records_text, simplifyVector = FALSE)

  if (as == "list") {
    return(records_json)
  }

  kntn_parse_records(records_json$records)
}

#' @rdname kntn_record
#' @export
kntn_records <- function(app, fields = NULL, query = "",
                         max_records = 1000L, offset = 0L, records_per_request = 100L,
                         as = c("data.frame", "list", "text"), verbose = FALSE) {
  if(!kntn_check_envvars_defined()) {
    stop("Please set necessary environmental variables by .Renviron or kntn_set_auth()")
  }

  url <- guess_url(Sys.getenv("KNTN_URL"))

  as <- match.arg(as)

  result <- list()

  # ignore if limit and offset is included in query
  if(!is.character(query) && is.na(query)) query <- ""
  query_base <- strip_query(query)

  # backdoor to boost request
  kntn_sleep_time <- Sys.getenv("KNTN_SLEEP_TIME")

  interval_time <- if(identical(kntn_sleep_time, "")) 1 else as.numeric(kntn_sleep_time)

  ranges <- calc_ranges(first_offset = offset, max_records = max_records, records_per_request = records_per_request)
  for (i in seq_len(nrow(ranges))) {
    limit  <- ranges$limits[i]
    offset <- ranges$offsets[i]
    query  <- paste(query_base, sprintf("limit %d offset %d", limit, offset))

    message(sprintf("Getting %d records from %d", limit, offset))
    result_tmp <- kntn_records_once(url = url, app = app, fields = fields,
                                     query = query, as = as, verbose = verbose)
    result[[i]] <- result_tmp

    # TODO: as="text"
    if(!is.character(result_tmp) && NROW(result_tmp) < limit) break

    Sys.sleep(interval_time)
  }

  if(as == "data.frame") dplyr::bind_rows(result) else result
}

kntn_stop_for_status <- function(res) {
  # Only HTTP status code 200 is success. Any other status codes are errors.
  code <- httr::status_code(res)
  if(identical(code, 200L)) {
    return(TRUE)
  }

  error_msg <- httr::content(res)$message %||% "(not available)"

  stop(sprintf("
  Status code: %d
  Error message: %s

If the error message is about authorization, please run kntn_set_auth(overwrite = TRUE)",
               code, error_msg))
}

strip_query <- function(x) {
  x <- stringr::str_replace(x, "limit\\s+\\d+(?=\\s+|$)", "")
  x <- stringr::str_replace(x, "offset\\s+\\d+(?=\\s+|$)", "")
  x
}
